server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,269 @@
+import org.infinispan.commons.dataconversion.MediaType;
+import org.infinispan.commons.dataconversion.Wrapper;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.encoding.DataConversion;
+import org.infinispan.multimap.impl.EmbeddedMultimapListCache;
+import org.infinispan.multimap.impl.ListBucket;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.ListenerHolder;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryCreated;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryModified;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+import org.infinispan.server.resp.Consumers;
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.SubscriberHandler;
+import org.infinispan.server.resp.commands.ArgumentUtils;
+import org.infinispan.server.resp.commands.Resp3Command;
+import org.infinispan.server.resp.filter.EventListenerKeysFilter;
+import org.infinispan.server.resp.logging.Log;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * @link https://redis.io/commands/blpop/
+ * @since 15.0
+ */
+public class BLPOP extends RespCommand implements Resp3Command {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass(), Log.class);
+
+   public BLPOP() {
+      super(-3, 1, -2, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
+         ChannelHandlerContext ctx,
+         List<byte[]> arguments) {
+      // Using last arg as timeout if it can be a double
+      int lastKeyIdx = arguments.size() - 1;
+      var timeout = ArgumentUtils.toDouble(arguments.get(arguments.size() - 1));
+      EmbeddedMultimapListCache<byte[], byte[]> listMultimap = handler.getListMultimap();
+      // If all the keys are empty or null, create a listener for each key
+      // otherwise return the left value of the first non empty list
+      var pollStage = pollKeyValue(listMultimap, arguments.get(0));
+      for (int i = 1; i < lastKeyIdx; ++i) {
+         var keyChannel = arguments.get(i);
+         // chain all polls returning first not null value or null/empty list
+         pollStage = pollStage.thenCompose(
+               v -> (v == null || v.isEmpty())
+                     ? pollKeyValue(listMultimap, keyChannel)
+                     : CompletableFuture.completedFuture(v));
+      }
+      // If no value returned, we need subscribers
+      return pollStage.thenCompose(v -> handler.stageToReturn(
+            (v != null && !v.isEmpty())
+                  ? CompletableFuture.completedFuture(v)
+                  : addSubscriber(arguments, lastKeyIdx, handler, ctx),
```

Comment:
After the subscription, we need to execute a verification with the pool again. There is a small interval after the checking and before adding the listener in which the entries might be created, and we don't receive an event.

Reply:
I've added some lines for that